monai/losses/image_dissimilarity.py

```diff
@@ -0,0 +1,155 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Union
+
+import torch
+from torch.nn import functional as F
+from torch.nn.modules.loss import _Loss
+
+from monai.networks.layers import gaussian_1d, separable_filtering
+from monai.utils import LossReduction
+
+
+class LocalNormalizedCrossCorrelationLoss(_Loss):
+    """
+    Local squared zero-normalized cross-correlation.
+    The loss is based on a moving kernel/window over the y_true/y_pred,
+    within the window the square of zncc is calculated.
+    The kernel can be a rectangular / triangular / gaussian window.
+    The final loss is the averaged loss over all windows.
+
+    Adapted from:
+        https://github.com/voxelmorph/voxelmorph/blob/legacy/src/losses.py
+        DeepReg (https://github.com/DeepRegNet/DeepReg)
+    """
+
+    def __init__(
+        self,
+        in_channels: int,
+        ndim: int = 3,
+        kernel_size: int = 9,
+        kernel_type: str = "rectangular",
+        reduction: Union[LossReduction, str] = LossReduction.MEAN,
+        smooth_nr: float = 1e-7,
+        smooth_dr: float = 1e-7,
+    ) -> None:
+        """
+        Args:
+            in_channels: number of input channels
+            ndim: number of spatial ndimensions, {``1``, ``2``, ``3``}. Defaults to 3.
+            kernel_size: kernel spatial size, must be odd.
+            kernel_type: {``"rectangular"``, ``"triangular"``, ``"gaussian"``}. Defaults to ``"rectangular"``.
+            reduction: {``"none"``, ``"mean"``, ``"sum"``}
+                Specifies the reduction to apply to the output. Defaults to ``"mean"``.
+
+                - ``"none"``: no reduction will be applied.
+                - ``"mean"``: the sum of the output will be divided by the number of elements in the output.
+                - ``"sum"``: the output will be summed.
+            smooth_nr: a small constant added to the numerator to avoid zero.
+            smooth_dr: a small constant added to the denominator to avoid nan.
+        """
+        super(LocalNormalizedCrossCorrelationLoss, self).__init__(reduction=LossReduction(reduction).value)
+        self.in_channels = in_channels
+
+        self.ndim = ndim
+        if self.ndim not in [1, 2, 3]:
+            raise ValueError(f"Unsupported ndim: {self.ndim}-d, only 1-d, 2-d, and 3-d inputs are supported")
+
+        self.kernel_size = kernel_size
+        if self.kernel_size % 2 == 0:
+            raise ValueError(f"kernel_size must be odd, got {self.kernel_size}")
+
+        if kernel_type == "rectangular":
```

Comment:
I suggest to keep a dictionary as a class attribute and use dict instead of if/else. similar to https://github.com/DeepRegNet/DeepReg/blob/main/deepreg/model/loss/image.py#L241.

Or maybe MONAI has different requirements ^^

Reply:
Done

Reply:
monai extends `Enum` https://github.com/Project-MONAI/MONAI/blob/a4ef69166c88ab252ef06268bc4c8fc7a089e36b/monai/utils/enums.py#L34 but this style is not yet enforced